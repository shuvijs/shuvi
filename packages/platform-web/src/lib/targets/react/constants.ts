export const BUILD_CLIENT_RUNTIME_REACT_REFRESH = `static/runtime/react-refresh`;
