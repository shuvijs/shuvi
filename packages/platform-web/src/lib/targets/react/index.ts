import { createPlugin } from '@shuvi/service';
import bundlerPlugin from './bundler';
import { resolveAppFile, resolveDep, resolveLib } from '../../paths';

const webReactMainPlugin = createPlugin({
  addRuntimeService: () => [
    {
      source: resolveAppFile('react/head/head'),
      exported: '{ default as Head }'
    },
    {
      source: resolveAppFile('react/dynamic'),
      exported: '{ default as dynamic }'
    },
    {
      source: resolveAppFile('react/loader'),
      exported: '{ useLoaderData }'
    },
    // typescript type cannot be exported at js file so put it at `index.d.ts`
    {
      source: resolveAppFile('react/loader'),
      exported: '{ Loader }',
      filepath: 'index.d.ts'
    },
    {
      source: resolveLib('@shuvi/router-react'),
      exported: '{ useParams, useRouter, useCurrentRoute, RouterView }'
    },
    {
      source: resolveAppFile('react/Link'),
      exported: '{ Link }'
    }
  ]
});
const platformWebReact = () => {
  return {
    plugins: [webReactMainPlugin, bundlerPlugin],
    platformModule: resolveAppFile('react/index'),
    polyfills: [
      resolveDep('react-app-polyfill/ie11'),
      resolveDep('react-app-polyfill/stable')
    ]
  };
};

export default platformWebReact;
