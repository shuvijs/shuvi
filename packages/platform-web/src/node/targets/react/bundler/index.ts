import * as path from 'path';
import { CorePluginConstructor, createPlugin } from '@shuvi/service';
import ReactRefreshWebpackPlugin from '@next/react-refresh-utils/ReactRefreshWebpackPlugin';
import { resolveLocal } from '../../../paths';
import { BUNDLER_TARGET_CLIENT } from '../../../../shared';

const isDefined = <T>(value: T | undefined): value is T => Boolean(value);

const configWebpack: CorePluginConstructor['configWebpack'] = (
  config,
  { name, webpack },
  context
) => {
  const resolveUser = (m: string, sub?: string) => {
    const { rootDir } = context.paths;
    let userPkg: { dependencies: Record<string, string> } = {
      dependencies: {}
    };
    // Check if target dependency is declared at user's package.json
    try {
      userPkg = require(path.join(rootDir, `package.json`));
      if (!userPkg.dependencies) {
        userPkg.dependencies = {};
      }
    } catch {}

    if (m in userPkg.dependencies) {
      // Resolve path of target dependency from user root path
      try {
        const module = sub ? `${m}/${sub}` : m;
        return require.resolve(module, { paths: [rootDir] });
      } catch {}
    }
    return undefined;
  };

  const isReactVersionAfter18 = () => {
    let version: string = '';
    try {
      const userReactDomPkgPath = resolveUser('react-dom', 'package.json');
      if (userReactDomPkgPath) {
        version = require(userReactDomPkgPath).version;
      }
    } catch {}
    if (!version) {
      version = require(resolveLocal('react-dom')).version;
    }
    const majorVersion = parseInt(version.split('.')[0] || '', 10);

    return majorVersion >= 18;
  };

  config.resolve.alias.set('@shuvi/service', resolveLocal('@shuvi/service'));
  config.resolve.alias.set('@shuvi/router$', resolveLocal('@shuvi/router'));
  config.resolve.alias.set(
    '@shuvi/router-react$',
    resolveLocal('@shuvi/router-react')
  );
  config.resolve.alias.set(
    'react$',
    // @ts-ignore
    [resolveUser('react'), resolveLocal('react')].filter(isDefined)
  );
  config.resolve.alias.set(
    'react/jsx-runtime$',
    // @ts-ignore
    [
      resolveUser('react', 'jsx-runtime'),
      resolveLocal('react', 'jsx-runtime')
    ].filter(isDefined)
  );
  config.resolve.alias.set(
    'react/jsx-dev-runtime$',
    // @ts-ignore
    [
      resolveUser('react', 'jsx-dev-runtime'),
      resolveLocal('react', 'jsx-dev-runtime')
    ].filter(isDefined)
  );
  config.resolve.alias.set(
    'react-dom$',
    // @ts-ignore
    [resolveUser('react-dom'), resolveLocal('react-dom')].filter(isDefined)
  );

  if (name === BUNDLER_TARGET_CLIENT) {
    config.plugin('version-env-plugin').use(webpack.DefinePlugin, [
      {
        'process.env.__SHUVI__AFTER__REACT__18__': JSON.stringify(
          isReactVersionAfter18()
        )
      }
    ]);

    if (context.mode === 'development') {
      config
        .plugin('react-refresh-plugin')
        .use(ReactRefreshWebpackPlugin, [webpack]);
    }
  }

  return config;
};

export default {
  core: createPlugin({
    configWebpack,
    addEntryCode(context) {
      if (context.mode === 'development') {
        const fastRefreshRuntime = require.resolve(
          `@next/react-refresh-utils/runtime`
        );
        return `import "${fastRefreshRuntime}"`;
      } else {
        return '';
      }
    }
  })
};
