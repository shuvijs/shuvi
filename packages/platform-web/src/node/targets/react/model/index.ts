import * as path from 'path';
import { createPluginAfter } from '@shuvi/service';
import { resolveLib, resolvePkgFile } from '../../../paths';

const core = createPluginAfter(
  {
    addRuntimeService: () => [
      {
        source: path.dirname(
          require.resolve('@shuvi/redox-react/package.json')
        ),
        exported: '*',
        filepath: 'model.ts'
      }
    ],
    configWebpack: config => {
      config.resolve.alias.set('@shuvi/redox', resolveLib('@shuvi/redox'));
      return config;
    }
  },
  {
    name: 'model-react'
  }
);

export const ModelReactPlugin = {
  core,
  runtime: {
    // this need
    plugin: resolvePkgFile('esm/shuvi-app/react/model/runtime')
  }
};
