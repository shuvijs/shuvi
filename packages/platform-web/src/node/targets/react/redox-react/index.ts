import * as path from 'path';
import { createPlugin } from '@shuvi/service';
import { resolveLib } from '../../../paths';

const runtimePath = require.resolve(
  path.join(
    __dirname,
    '../../../../../esm/node/targets/react/redox-react/runtime'
  )
);
const core = createPlugin({
  addRuntimeService: () => [
    {
      source: path.dirname(require.resolve('@shuvi/redox-react/package.json')),
      exported: '*',
      filepath: 'model.ts'
    }
  ],
  configWebpack: config => {
    config.resolve.alias.set('@shuvi/redox', resolveLib('@shuvi/redox'));
    return config;
  }
});
export const RedoxReactPlugin = {
  core,
  runtime: {
    plugin: runtimePath
  }
};
