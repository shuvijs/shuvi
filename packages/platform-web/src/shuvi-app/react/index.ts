import view from './view';
import app from './App';
import getRoutes from './getRoutes';
export { view, app, getRoutes };
