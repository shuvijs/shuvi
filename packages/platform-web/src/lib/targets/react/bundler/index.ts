import * as path from 'path';
import { resolveLocal as _resolveLocal } from '@shuvi/utils/lib/resolve';
import { CorePluginConstructor, createPlugin } from '@shuvi/service';
import ReactRefreshWebpackPlugin from '@next/react-refresh-utils/ReactRefreshWebpackPlugin';
import { BUNDLER_DEFAULT_TARGET } from '@shuvi/shared/lib/constants';
import { PACKAGE_DIR } from '../../../paths';
import { BUILD_CLIENT_RUNTIME_REACT_REFRESH } from '../constants';
import serverPlugin from './insertReactRefreshEntryFile';
import { DefinePlugin } from 'webpack';

const configWebpack: CorePluginConstructor['configWebpack'] = (
  config,
  { name, webpack },
  context
) => {
  const resolveLocal = (id: string) =>
    _resolveLocal(id, { basedir: __dirname });
  const resolveUser = (m: string) =>
    path.join(context.paths.rootDir, 'node_modules', m);

  const isReactVersionAfter18 = () => {
    let version: string = '';
    try {
      version = require(resolveUser('react-dom')).version;
    } catch (e) {
      version = require(resolveLocal('react-dom')).version;
    } finally {
    }
    const majorVersion = parseInt(version.split('.')[0] || '', 10);

    return majorVersion >= 18;
  };

  config.resolve.alias.set('@shuvi/platform-web', PACKAGE_DIR);
  config.resolve.alias.set('@shuvi/service', resolveLocal('@shuvi/service'));
  config.resolve.alias.set(
    '@shuvi/router-react$',
    resolveLocal('@shuvi/router-react')
  );
  config.resolve.alias.set(
    '@shuvi/redox-react$',
    resolveLocal('@shuvi/redox-react')
  );
  // @ts-ignore
  config.resolve.alias.set('react$', [
    resolveUser('react'),
    resolveLocal('react')
  ]);
  // @ts-ignore
  config.resolve.alias.set('react-dom$', [
    resolveUser('react-dom'),
    resolveLocal('react-dom')
  ]);

  if (name === BUNDLER_DEFAULT_TARGET && context.mode === 'development') {
    config.module
      .rule('main')
      .oneOf('js')
      .use('react-refresh-loader')
      .loader(require.resolve('@next/react-refresh-utils/loader'))
      .before('shuvi-swc-loader');

    config
      .plugin('react-refresh-plugin')
      .use(ReactRefreshWebpackPlugin, [webpack]);
    config.plugin('version-env-plugin').use(DefinePlugin, [
      {
        'process.env.__SHUVI__AFTER__REACT__18__': JSON.stringify(
          isReactVersionAfter18()
        )
      }
    ]);
    config.merge({
      entry: {
        [BUILD_CLIENT_RUNTIME_REACT_REFRESH]: [
          require.resolve(`@next/react-refresh-utils/runtime`)
        ]
      }
    });
  }
  return config;
};

export default {
  core: createPlugin({
    configWebpack
  }),
  server: serverPlugin
};
