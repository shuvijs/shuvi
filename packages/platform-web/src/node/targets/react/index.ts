import { createPlugin } from '@shuvi/service';
import bundlerPlugin from './bundler';
import { ModelReactPlugin } from './model';
import { resolvePkgFile, resolveDep } from '../../paths';

const webReactMainPlugin = createPlugin({
  addRuntimeService: () => [
    {
      source: resolvePkgFile('esm/shuvi-app/react/shuvi-runtime-api'),
      exported: '*'
    }
  ]
});
const platformWebReact = () => {
  return {
    plugins: [webReactMainPlugin, bundlerPlugin, ModelReactPlugin],
    platformModule: resolvePkgFile('esm/shuvi-app/react/index'),
    polyfills: [
      resolveDep('react-app-polyfill/ie11'),
      resolveDep('react-app-polyfill/stable')
    ]
  };
};

export default platformWebReact;
