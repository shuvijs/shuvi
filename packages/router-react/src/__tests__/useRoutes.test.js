import * as React from 'react';
import { create as createTestRenderer } from 'react-test-renderer';
import { MemoryRouter as Router, useRoutes } from 'react-router';

describe('useRoutes', () => {
  it('returns the matching element from a route config', () => {
    function RoutesRenderer({ routes }) {
      return useRoutes(routes);
    }

    function Home() {
      return <h1>Home</h1>;
    }

    function About() {
      return <h1>About</h1>;
    }

    let routes = [
      { path: 'home', element: <Home /> },
      { path: 'about', element: <About /> }
    ];

    let renderer = createTestRenderer(
      <Router initialEntries={['/home']}>
        <RoutesRenderer routes={routes} />
      </Router>
    );

    expect(renderer.toJSON()).toMatchSnapshot();
  });
});
