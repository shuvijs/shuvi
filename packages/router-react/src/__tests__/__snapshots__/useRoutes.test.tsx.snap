// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useRoutes returns the matching element from a route config 1`] = `
<h1>
  Home
</h1>
`;
